templates/install_puppet.ps1.erb

```diff
@@ -0,0 +1,185 @@
+# install_puppet.ps1
+<#
+.Synopsis
+  Install or upgrade puppet-agent
+.Description
+  This script will install or upgrade puppet-agent on a windows machine from the MSI file at $Source.
+  If the script is supplied with $PuppetPID it will wait on the $PuppetPID before attempting to perform
+  an upgrade.
+.Parameter PuppetPID
+  The process ID of puppet to wait on before executing an upgrade. Note that all puppet processes must shut
+  down before an installation can occur
+.Parameter Source
+  The location of the new puppet-agent MSI installer
+.Parameter Logfile
+  File location where the installation will log output
+.Parameter InstallDir
+  Optionally change the default location where puppet-agent will be installed
+.Parameter PuppetMaster
+  The location of the puppet master
+.Parameter PuppetStartType
+  Optionally change the default start type of puppet
+.Parameter InstallArgs
+  Provide any extra argmuments to the MSI installation
+.Parameter UseLockedFilesWorkaround
+  Set to $true to enable execution of the puppetres.dll move workaround. See https://tickets.puppetlabs.com/browse/MODULES-4207
+#>
+[CmdletBinding()]
+param(
+  # PuppetPID _must_ come first!, this script needs PuppetPID to be a positional parameter to execute
+  # correctly from the module.
+  [String] $PuppetPID,
+  [String] $Source='<%= @_msi_location %>',
+  [String] $Logfile='<%= @_logfile %>',
+  [String] $InstallDir='<%= @install_dir %>',
+  [String] $PuppetMaster='<%= @_puppet_master %>',
+  [String] $PuppetStartType='<%= @_agent_startup_mode %>',
+  [String] $InstallArgs='<%= @_install_options.map do |option|

```

Comment:
Personally I'd prefer this "logic" be in the manifest and the $InstallArgs be a du,b string, but I'm not too fussed

Reply:
I agree with Glenn